image_eval/pairwise_evaluators.py

```diff
@@ -0,0 +1,124 @@
+import torch
+
+from PIL import Image
+from torchmetrics import MultiScaleStructuralSimilarityIndexMeasure
+from torchmetrics import PeakSignalNoiseRatio
+from torchmetrics import UniversalImageQualityIndex
+from torchmetrics.image.lpip import LearnedPerceptualImagePatchSimilarity
+from torchvision import transforms
+from typing import Any, Dict, List
+
+from image_eval.evaluators import BaseEvaluator
+from image_eval.evaluators import EvaluatorType
+
+
+class PairwiseSimilarityEvaluator(BaseEvaluator):
+    """Evaluates how similar pairs of images are.
+
+    Helpful to compare against images generated from consecutive checkpoints during training. When
+    these metrics start signaling high similarity between consecutive checkpoints, it's a sign we
+    can stop training, since the model has converged.
+
+    All the children of these methods are metrics found in HEIM:
+    https://crfm.stanford.edu/helm/heim/latest/
+    """
+    TYPE = EvaluatorType.PAIRWISE_SIMILARITY
+
+    def __init__(self, device: str, metric_name: str, metric: Any, normalize: bool = False):
+        """Constructs a pairwise evaluator that uses a given metric from `torchmetrics`.
+
+        Args:
+            device: "cpu" or "cuda"
+            metric_name: A human-readable name for the metric
+            metric: An object that has a __call__(images1, images2) method for pairwise comparison
+            normalize: Whether images need to be normalized to [0, 1] before passing to the metric
+        """
+        super().__init__(device)
+        self.metric_name = metric_name
+        self.metric = metric.to(device)
+        self.normalize = normalize
+
+    def evaluate(self,
+                 generated_images: List[Image.Image],
+                 real_images: List[Image.Image],
+                 **unused_kwargs) -> Dict[str, float]:
+        if len(generated_images) != len(real_images):
+            raise ValueError("Pairwise evaluators expect 1:1 pairs of generated/real images.")
+        else:
+            import logging
+            logging.info(f"Comparing {len(generated_images)} pairs of images.")
+
+        img_transforms = [transforms.Resize((256, 256)), transforms.ToTensor()]
+        if self.normalize:
+            img_transforms.append(transforms.Normalize(mean=(0.5, 0.5, 0.5), std=(0.5, 0.5, 0.5)))
+
+        preprocessing = transforms.Compose(img_transforms)
+        generated_images = [preprocessing(img) for img in generated_images]
+        real_images = [preprocessing(img) for img in real_images]
+
+        generated_images = torch.stack(generated_images).to(self.device)
+        real_images = torch.stack(real_images).to(self.device)
+        logging.info("Mean delta:")
```

Comment:
nit: just use an f-string and then you can combine the two logging.info calls

Reply:
I'll actually remove the logging statements, I left them in by mistake.